core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt

```diff
@@ -59,6 +59,36 @@ public interface SupportedDataFrameFormat : SupportedFormat {
     public fun readDataFrame(file: File, header: List<String> = emptyList()): DataFrame<*>
 }
 
+/**
+ * User-facing API implemented by a companion object of an imported schema [org.jetbrains.kotlinx.dataframe.annotations.DataSchemaSource]
+ * Intended use:
+ * 1. Interact with this API to read dataframe of a desired type
+ * 2. API is used to implement "generic dataframe reader" based on [schemaKType] and other available methods
+ */
+public interface DataFrameProvider<T> {
+    public val schemaKType: KType
+
+    public fun default(): DataFrame<T>
+
+    public fun read(path: String): DataFrame<T>
+}
+
+/**
+ * Handler of classes annotated with [org.jetbrains.kotlinx.dataframe.annotations.DataSchemaSource].
+ * Implementations must have a single zero-argument constructor
```

Comment:
they could also be `object` singletons maybe, since they have no state

Reply:
Good idea! It first needs to be adjusted in compiler plugin